test/state-component-mocha.js

```diff
@@ -0,0 +1,288 @@
+
+var _ = require('underscore');
+var noflo = require('noflo');
+var stateComponent = require('../components/state-component.js');
+
+describe('state-component', function() {
+    it("should reject undefined definition", function() {
+        return Promise.resolve().then(stateComponent).should.be.rejected;
+    });
+    it("should reject empty definition", function() {
+        return Promise.resolve({}).then(stateComponent).should.be.rejected;
+    });
+    it("should reject empty inPorts", function() {
+        return Promise.resolve({inPorts:{},onchange:function(){}}).then(stateComponent).should.be.rejected;
+    });
+    it("should reject no onchange function", function() {
+        return Promise.resolve({
+            inPorts:['in']
+        }).then(stateComponent).then(createComponent).then(function(component){
+            return _.keys(component.inPorts);
+        }).should.be.rejected;
+    });
+    it("should accept array of inPort names", function() {
+        return Promise.resolve({
+            inPorts:['in'],
+            onchange:function(){}
+        }).then(stateComponent).then(createComponent).then(function(component){
+            return _.keys(component.inPorts);
+        }).should.eventually.contain('in');
+    });
+    it("should trigger onchange function", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:['hello'],
+            onchange: function(state) {
+                handler(state);
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'hello', "world");
+            });
+        }).should.become({'hello': "world"});
+    });
+    it("should pass previous resolve state as second argument", function() {
+        var handler;
+        return Promise.resolve({
+            inPorts:['hello'],
+            onchange: function(state, previously) {
+                handler(previously);
+                return state.hello;
+            }
+        }).then(stateComponent).then(createComponent).then(function(component){
+            // have the handler call a Promise resolve function to
+            // check that the data sent on the in port is passed to the handler
+            return new Promise(function(callback){
+                handler = callback;
+                sendData(component, 'hello', "world");
+            }).then(function(){
+                return new Promise(function(callback){
+                    handler = callback;
+                    sendData(component, 'hello', "again");
+                });
+            });
+        }).should.become("world");
+    });
+    it("shouldn't confuse an undefined key with non-indexed port", function() {
```

Comment:
How about something like

shouldn't confuse data with an undefined key on indexed port with incoming data on a non-indexed port

And add something somewhere that explains the difference between an indexed and wildcarded port.